### DISCUSSION TITLE
SMTP Server

### DISCUSSION THREAD

**Discussion by MrJohnson-SixNinesIT** [lpm0073/cookiecutter-openedx-devops#20] [Q&A] [answered]
Howdy,

How is the mail server functionality intended to work? Does all it require is that those email addresses exist, and the MX records for the external host (I.E. something like GoDaddy for) are copied into the primary hosted zone in Route53?

If the above is true, would adding the records in Route53 and running the Github Actions update things accordingly?

Lastly, in the deploy-{env}.yml there is the “Configure SMTP” step. In the “Configure AWS credentials” step, we were able to use our session token to authenticate and get things working. In the “Configure SMTP” step, the session token is not recognized and shows a warning. Is the session token variable required in this step? If we exclude attempting to use a session token, we do NOT see the warning. Wondering if this will cause an issue with the emails down the line.

An example of our credentials step that works (in our case it will NOT work without it):

![image](https://user-images.githubusercontent.com/105312312/188724485-cf856f79-cb0d-413b-9f39-5414145843a0.png)

Here is the warning we receive when we attempt to use the session token in the “Configure SMTP” step:

![image](https://user-images.githubusercontent.com/105312312/188724562-0255aeda-7362-4146-93a3-d4b01b1cbc57.png)

**Comment by lpm0073** [comment]
there's a bit of apples-and-oranges in your question, so i'll respond in pieces to try to unwind.

**configuring SMTP**: the short answer is that you only need to set 4 Python variables in the LMS:
 - https://github.com/openedx/edx-platform/blob/master/lms/envs/production.py#L151
 - https://github.com/openedx/edx-platform/blob/master/lms/envs/production.py#L495
there potentially are additional details,  like say tls and ssl, depending on your email provider. I prefer AWS SES but also occasionally use Google mail services at clients' requests. The hook that i provide in the deployment work flow is simply some scaffolding to grab the username and password values from Github Actions secrets. please see the README in your openedx_devops repo for more info, and screen shots.

more ....

aws-actions/configure-aws-credentials: the purpose of this Github Action, created and maintained by AWS, is to configure AWS credential and region environment variables for use in other GitHub Actions. The environment variables will be detected by both the AWS SDKs and the AWS CLI to determine the credentials and region to use for AWS API calls. Thus, it generates a session token based on an AWS IAM key pair that you provide. hence, the framing of your question above is at odds w what this Action actually does. More generally, and to your broader question about configuring SMTP: this AWS Action is unrelated to configuring SMTP for open edx. you can read more here: https://github.com/marketplace/actions/configure-aws-credentials-action-for-github-actions

hope that helps.

**Reply by MrJohnson-SixNinesIT** [reply]
Okay, so that makes sense. 

So the host and port look fine. We added the EMAIL_HOST_PASSWORD and EMAIL_HOST_USER secrets to the workflow, but it does not appear to be setting the EMAIL_HOST_USER.

![image](https://user-images.githubusercontent.com/105312312/189160556-6daa13e3-c7b6-40a2-9393-7e25731c0b47.png)

When looking at the workflow, I don't see it there in the list of variables at all:

![image](https://user-images.githubusercontent.com/105312312/189161185-a074499e-039a-46c4-9a72-666b90800adb.png)

So I guess my question is, where does the value go  for the EMAIL_HOST_USER, if not the Actions Secrets?

**Reply by xavierperez06** [reply]
Hello, i'm having the same problem. I've deployed everything and it seems to be working fine except the email. If i go the AWS SES i don't see anything configured there, should i have anything in there, or do any extra step?

Also, regarding the 4 Python variables for the LMS you @lpm0073 you pointed out, those variables should be in the file **settings_merge.yml**? Because i don't see EMAIL_HOST or EMAIL_HOST_PASSWORD anywhere.

**Comment by lpm0073** [comment]
i need to improve the documentation. responding to questions from you both:
- the short answer is that you either set a bash environment variable (does not matter how you do this) of the form TUTOR_EMAIL_HOST_PASSWORD, which tutor will then consume and use on the next "tutor config save" operation, or, you explicitly set the value using a call such as "tutor config save --set EMAIL_HOST_PASSWORD='YADDAYADDA'". The Github Actions that i wrote do this for you, but, you can also create these variables yourself.
- you can read values from Actions Secrets, but you cannot write to them. You set Actions Secrets from the screen in the screen shot, below.
<img width="1167" alt="Screen Shot 2022-09-11 at 18 55 16" src="https://user-images.githubusercontent.com/1178819/189554389-52d84cc3-48da-4b52-8030-11e7ec71c987.png">

**Comment by MrJohnson-SixNinesIT** [comment]
Hello, so we were able to get credentials to work by passing in the values into the tutor config save command:

![image](https://user-images.githubusercontent.com/105312312/189972117-586fc0fb-b710-4eba-ae09-8d4ac8686cff.png)


Although this appears to work, emails are not being sent out. We are using AWS Simple Email Service with the following configurations:

In SES:

- Domain Identity has been verified.
- Email addresses such as “support@domain”, “registration@domain” etc… have all been verified.
- The email host is the is SMTP endpoint value value in SES.

We’ve confirmed in SES that emails could be sent out. And just to be sure, we registered at courses.{ourDomainName}.com using an SES verified email, but still did not see any emails go out from SES.

I feel like there is something else we are missing here, but not quite sure what that is.

**Reply by lpm0073** [reply]
Glad you're progressing. A couple of comments:

1. Refer to this code example to read your AWS SES credentials after you've moved these to Github Secrets: [https://github.com/StepwiseMath/openedx_devops/](https://github.com/StepwiseMath/openedx_devops/blob/main/.github/workflows/deploy-stepwise-prod.yml#L49)

2. Keep in mind that SMTP email is just a back end service, same as any other like mysql, mongo, redis, etcetera. Open edX software simply uses whatever connection scheme and credentials you provide. For testing email i mostly rely on the password reset workflow, mainly because i know that Open edX immediately attempts to send a reset email whereas with other email work flows i'm never 100% certain when it will attempt to do this.

**Reply by MrJohnson-SixNinesIT** [reply]
@lpm0073 that repository may be private.

**Reply by lpm0073** [reply]
Here are the relevant screen shots. 

**Most importantly**, please note that i noticed a syntactical error in the casing of the key pair secrets for AWS SES. note the two edited rows in the screen shot below. This clearly is why you were struggling. I'll modify the code snippet in this cookiecutter accordingly.

<img width="670" alt="Screen Shot 2022-09-14 at 7 22 34" src="https://user-images.githubusercontent.com/1178819/190152294-92cab15e-9239-49dc-a1cf-5f1bfa28ef68.png">

**Reply by MrJohnson-SixNinesIT** [reply]
@lpm0073 Oh, for our code, we had caught that and have been using the upper-casing in our code, much like your screenshots depict:

![image](https://user-images.githubusercontent.com/105312312/190173849-1e194b62-6972-4b58-829e-22b2da95b4bc.png)

Per your suggestion, we have also started using the password reset functionality for testing, and do not see any emails being sent in SES. The IAM user that we have configured for this also has SendRawEmail permissions.

And for testing purposes, we have even hardcoded the values for the aws-ses-id and secret, but still do not see any emails going out.

Is there some way to see logs or something as to why this would be failing to send?

**Reply by lpm0073** [reply]
yes, you can view the lms logs directly from inside of k9s by using the "L" hotkey on an lms pod. alternatively, there's a cookiecutter command-line script available in the newest version of the ec2_bastion -- `klog.sh` that will download pod log data to your console.

Either way, keep in mind that the email processing will be sent to a random lms pod and so you'll need to keep an eye on all lms pods that are running. 

With all of that in mind, i'm usually able to see and diagnose all SMTP send problems from the lms log data. 

<img width="757" alt="Screen Shot 2022-09-14 at 12 59 09" src="https://user-images.githubusercontent.com/1178819/190228349-61d4ab15-e27f-4f44-99bf-1cd2db17ef4b.png">

**Reply by MrJohnson-SixNinesIT** [reply]
Okay, so thanks for the awesome advice on using k9s to view the logs. I wasn't able to see anything pertinent to this on the LMS pods, but I was able to view logs on the SMTP server. 

On the SMTP server, we get timeouts: 

![image](https://user-images.githubusercontent.com/105312312/190474135-f71a2010-8d84-494d-8187-a3053f28192a.png)


Just to be sure, we tested sending emails using AWS CLI and that worked without issue. Sending emails worked from the Sandbox as well. 

Just in case, we also tried different email providers, and we still never see emails being sent.

**Reply by lpm0073** [reply]
The following is a general purpose smart bomb for trouble shooting purposes. you can use the following from inside the shell of an lms pod to dump EVERY django setting, including any and all SMTP email settings:

`./manage.py lms --settings=tutor.production diffsettings --all`

**Reply by MrJohnson-SixNinesIT** [reply]
@lpm0073 Okay, we were able to dig further into the logs and found the following:

![image](https://user-images.githubusercontent.com/105312312/191100637-229c0186-c586-4b94-bcc8-7bb2b4b03845.png)

Above, we get the error of "SMTP AUTH extension not supported by server". 

With the following settings being passed in:

![image](https://user-images.githubusercontent.com/105312312/191100788-d2f00693-87b6-4520-a41a-11068dc69559.png)

**Comment by MrJohnson-SixNinesIT** [comment]
When we look online about the "SMTP AUTH extension not supported by server" error, people who receive that message in other contexts mention a port issue, but port 587 seems to work for them.

**Reply by lpm0073** [reply]
any update on this?

**Reply by MrJohnson-SixNinesIT** [reply]
@lpm0073 unfortunately we are still unable to figure out why the emails don't send.

**Reply by lpm0073** [reply]
ok. i have to deploy a large platform this coming saturday. so, one way or another i'll figure this out.

**Reply by MrJohnson-SixNinesIT** [reply]
@lpm0073 Awesome! Thank you. Looking forward to seeing what you turn up.

**Comment by MrJohnson-SixNinesIT** [comment]
@lpm0073 Howdy. Did you have any luck this past weekend?

**Comment by MrJohnson-SixNinesIT** [comment]
I'm wondering if this functionality can be overridden with a plug perhaps. Or if that isn't the intended purpose of the plugins.

**Comment by MrJohnson-SixNinesIT** [comment, marked as answer]
@lpm0073 Ultimately, we got things working by adding those port, email_host_user and email_host_password directly to the settings_merg.yml. 

Once that worked, we went back and created secrets with those same names and that got email working for us via AWS SES.
